plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperations.java

```diff
@@ -102,13 +104,36 @@ protected String getRefreshedLocation(boolean invalidateCaches)
     protected void commitNewTable(TableMetadata metadata)
     {
         verify(version.isEmpty(), "commitNewTable called on a table which already exists");
+        CommitStatus status = CommitStatus.FAILURE;
         String newMetadataLocation = writeNewMetadata(metadata, 0);
         TableInput tableInput = getTableInput(tableName, owner, ImmutableMap.of(METADATA_LOCATION_PROP, newMetadataLocation));
 
         CreateTableRequest createTableRequest = new CreateTableRequest()
                 .withDatabaseName(database)
                 .withTableInput(tableInput);
-        stats.getCreateTable().call(() -> glueClient.createTable(createTableRequest));
+        try {
+            stats.getCreateTable().call(() -> glueClient.createTable(createTableRequest));
+            status = CommitStatus.SUCCESS;
+        }
+        catch (AmazonServiceException e) {
+            if (e instanceof OperationTimeoutException) {
+                status = CommitStatus.UNKNOWN;
+            }
+            throw e;
+        }
+        catch (RuntimeException e) {
+            // Check whether the creation of the table eventually succeeded
+            status = checkCommitStatus(newMetadataLocation);
+            switch (status) {
+                case FAILURE -> throw e;
+                case UNKNOWN -> throw new CommitStateUnknownException(e);
+            }
+        }
+        finally {
+            if (status == CommitStatus.FAILURE) {
+                fileIo.deleteFile(newMetadataLocation);
```

Comment:
Looks like we are reaching here on any `AmazonServiceException` that is not `OperationTimeoutException`.
What if this is some other exception that prevented us from knowing whether a call was successful? Like hypothetical NetworkWasSuddenDisconnectedException?

I think we should rather have an allow-list of exceptions that indicate known failure. Permission denied, no such object, etc -- failures indicating lack of action clearly.

Reply:
Also, do we need to delete the file explicitly?
if we do **not** throw CommitStateUnknownException, will it get removed by the library?

(i am not a fan of implicitness of CommitStateUnknownException handling, but since it exists anyway in the library, it's ok to leverage it where it works for us)

Reply:
> Also, do we need to delete the file explicitly?

Yes, we have to delete the metadata file explicitly. The metadata file is created inside `commitNewTable` and the file is not tracked yet by iceberg library.

>if we do not throw CommitStateUnknownException, will it get removed by the library?

No, It won't. the library is responsible for deleting the snapshot files in case of failure, but not metadata file.